docs/development/systemtask.md

```diff
@@ -0,0 +1,263 @@
+
+Development environment
+-----------------------
+
+You should have the
+[crazyflie-firmware](https://github.com/bitcraze/crazyflie-firmware) and
+[crazyflie-clients-python](https://github.com/bitcraze/crazyflie-clients-python)
+cloned in the same folder. If you are using the [Bitcraze
+VM](https://github.com/bitcraze/bitcraze-vm) this is already the case.
+
+The Crazyflie firmware should be on Master branch.
+
+For the rest of the howto you will work in the crazyflie-firmware
+project.
+
+
+Implementing the task
+---------------------
+
+High level tasks (those that do not directly talk to hardware) usually go in
+the `modules` directory. We will walk through the sections needed in a
+new file `modules/src/example.c`.
+
+
+First, include the necessary system header files:
+
+``` {.c}
+#include "FreeRTOS.h"
+#include "queue.h"
+#include "static_mem.h"
+#include "task.h"
+```
+
+### File-static variables
+
+Next, we statically allocate the queue to hold inputs from
+other parts of the system such as the radio.
+Here we just hold integers in the queue. Often it will be a struct instead.
+A queue of length 1 is appropriate if new inputs supersede old ones.
+For example, if the input controls the current LED color,
+we do not care which other colors were requested in the past.
+
+Note that the `STATIC_MEM_*_ALLOC` family of macros expand to static
+variable declarations, and can be used outside of any function scope.
+
+``` {.c}
+static xQueueHandle inputQueue;
+STATIC_MEM_QUEUE_ALLOC(inputQueue, 1, sizeof(int));
+```
+
+
+Next, we allocate the task call stack and other necessary memory for OS
+bookkeeping. Note that `configMINIMAL_STACK_SIZE` gives a fairly small stack.
+Tasks that do lots of computation will almost certainly need larger stacks.
+
+The FreeRTOS task API requires that task functions take a `void *` argument.
+
+``` {.c}
+static void exampleTask(void*);
+STATIC_MEM_TASK_ALLOC(exampleTask, configMINIMAL_STACK_SIZE);
+```
+
+### Functions
+
+Next, we implement the `Init()` function.
+This should contain all of the FreeRTOS allocations and initializations
+that need to last as long as the firmware is running.
+Later in this howto, we will need to edit another file to call the `Init()`.
+
+``` {.c}
+static bool isInit = false;
+
+void exampleTaskInit() {
+  inputQueue = STATIC_MEM_QUEUE_CREATE(inputQueue);
+  // TODO
+  STATIC_MEM_TASK_CREATE(exampleTask, exampleTask, KALMAN_TASK_NAME, NULL, KALMAN_TASK_PRI);
+  isInit = true;
+}
+```
+
+Next, we implement the `Test()` function.
+Usually this does not do much, besides verify that `Init()` has been called.
+Later in this howto, we will need to edit another file to call the `Test()`.
+
+``` {.c}
+bool exampleTaskTest() {
+  return isInit;
+}
+```
+
+
+Now we come to the main task function.
+It is usually an infinite loop that waits on inputs from something.
+In our case, we wait on inputs from the queue.
+
+Our `xQueueReceive` call waits for up to 1000 ticks - one second.
```

Comment:
It might be a good idea to mention that ```portMAX_DELAY``` will wait forever and only release the lock when the queue is not empty. I think this is a common use case.

Reply:
Switched from 1000ms to `portMAX_DELAY` timeout value and added explanation.